Feature rapport final (#8)
CQEN-QDCE/exp-dev-distribue: exp/rapport.md

Des services aux développeurs sous forme d’API ont donc été ébauchés pour imaginer quels outils seraient nécessaire pour supporter un processus de publication libre-service.  

![Diagramme de conteneurs](https://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://raw.githubusercontent.com/CQEN-QDCE/exp-dev-distribue/main/exp/architecture/conteneurs.puml)

Le contributeur de service interagit avec le Registre des services. Des services d’approbation et de publication sous forme d’API REST permettent d’insérer ceux-ci dans un pipeline de déploiement chez le contributeur.   

![Diagramme de flux](https://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://raw.githubusercontent.com/CQEN-QDCE/exp-dev-distribue/main/exp/architecture/fluxPublication.puml)

Le processus de publication obligerait de passer l’étape d’approbation avec succès avant d’invoquer le service de publication. Ce dernier pourrait être invoqué par l’orchestration du déploiement du service du contributeur par exemple. 

> En mettant ainsi à profit le modèle de communauté ouverte en libre-service, on vient assurer la prise en charge des développeurs avec un minimum de ressource humaine. Un gestionnaire de communauté peut assurer les interactions avec les développeurs de services, tandis que les mainteneurs du produit prennent connaissance des problématiques directement dans les outils de rapport de bogue. Le reste des interactions sont automatisés grâce à mise en disponibilité d’outils libre-service. 

Reviewer: ```suggestion
En mettant ainsi à profit le modèle de communauté ouverte en libre-service, on vient assurer la prise en charge des développeurs avec un minimum de ressource humaine. Un gestionnaire de communauté peut assurer les interactions avec les développeurs de services, tandis que les mainteneurs du produit prennent connaissance des problématiques directement dans les outils de rapport de bogue. Le reste des interactions sont automatisés grâce à la mise en disponibilité d’outils libre-service. 
```

    Les couches d’accès multi-plateformes (*Cross-platform Native Runtime*) sont des librairies de codes utilisables dans des applications web permettant d’accéder aux fonctionnalités natives des systèmes mobiles combinées à un outil d’empaquetage pouvant construire des applications natives. L’utilisation de ces librairies permet de prendre à peu près n’importe quel langage ou framework web pour construire une application multi-plateforme distribuable dans les *app stores* ou exécutable dans un navigateur. 

    L’utilisation de ces librairies permet de prendre une approche de développement orientée d’abord pour le web et la rendre mobile sans trop de compromis. On s’assure ici une très forte compatibilité avec les standards du web, tel que les *web components*, ce qui se traduit par de meilleures performances, une accessibilité rehaussée ainsi qu’une meilleure affinité avec les algorithmes de classement des moteurs de recherche. 

    Les principales librairies de couche d’accès multi-plateforme sont [Capacitorjs](https://capacitorjs.com/docs), [Nativescript](https://docs.nativescript.org/) et [Cordova](https://cordova.apache.org/docs/en/latest/).  

* **Frameworks hybrides**

>     Les frameworks de développement hybrides ont été créé précisément pour concevoir des applications distribuables à la fois sur mobile sur le web. Ils utilisent dans la plupart des cas des langages de programmation qui ne sont pas basés sur l’écosystème javascript et utilisent un processus de compilation/transpilation pour créer un exécutable compatible avec la plateforme voulue.  

Reviewer: ```suggestion
    Les frameworks de développement hybrides ont été créé précisément pour concevoir des applications distribuables à la fois sur mobile et sur le web. Ils utilisent dans la plupart des cas des langages de programmation qui ne sont pas basés sur l’écosystème javascript et utilisent un processus de compilation/transpilation pour créer un exécutable compatible avec la plateforme voulue.  
```

    Leurs cas d’utilisations principaux est le développement applicatif avec interfaces qui se rapprochent le plus possible d’une expérience native. Cette spécialisation fait que ces frameworks se prêtent moins bien au développement de site ou de portail web. En effet, le rendu web (HTML/CSS/JS) étant le résultat d’une transpilation, il est difficile de contrôler le rendu final du portail et ainsi d’assurer une bonne compatibilité avec les standards du web ou encore d’assurer l’affinité avec les algorithmes de classement des moteurs de recherche.  

>     Les principaux frameworks de développement hybride sont[Flutter (Google)](https://flutter.dev/), [Blazor (Microsoft)](https://dotnet.microsoft.com/en-us/apps/aspnet/web-apps/blazor) et [React Native for Web (Meta)](https://necolas.github.io/react-native-web/)	 

Reviewer: ```suggestion
    Les principaux frameworks de développement hybride sont: [Flutter (Google)](https://flutter.dev/), [Blazor (Microsoft)](https://dotnet.microsoft.com/en-us/apps/aspnet/web-apps/blazor) et [React Native for Web (Meta)](https://necolas.github.io/react-native-web/)	 
```

L'expérimentation a arrêté son choix sur le framework Angular pour implémenter un premier exemple de service sous forme de web component. Angular supporte la transpilation de ses composants en *custom element* standard via son *package* [elements](https://angular.io/guide/elements). Celui-ci fourni une fonction qui donne l'instruction de déclarer un composant sous forme de *custom element*.

Dans cet exemple (`main.ts`) on démontre la simplicité de définir le composant `AppComponent` en tant que *custom element* `mea-service-exemple`. 
```typescript
import { createApplication } from '@angular/platform-browser';
import { createCustomElement } from '@angular/elements';
import { appConfig } from './app/app.config';
import { AppComponent } from './app/app.component';

(async () => {

    const app = await createApplication(appConfig);

    const appElement = createCustomElement(AppComponent, {
      injector: app.injector,
    });

    customElements.define('mea-service-exemple', appElement);
  
})();
```

> On peut donc développer une application Angular selon les patrons habituels et inclure celle-ci en tant que *microfrontend*. Par contre l'expérimentation a permis de découvrir quelques particularités auxquels il faire attention lorsqu'on développe un composant dans le but de le distribuer vers notre application coquille.

Reviewer: ```suggestion
On peut donc développer une application Angular selon les patrons habituels et inclure celle-ci en tant que *microfrontend*. Par contre, l'expérimentation a permis de découvrir quelques particularités auxquelles il faut faire attention lorsqu'on développe un composant dans le but de le distribuer vers notre application coquille.
```

L'exemple suivant montre comment naviguer dans le composant Angular sans interaction avec la barre d'adresse en ajoutant le [paramètre de configuration `skipLocationChange`](https://angular.io/api/router/NavigationBehaviorOptions#skipLocationChange) aux appels de la fonction `navigate`:
```typescript
export class FormulaireComponent {
    constructor(
        private router:Router
    ){}

    exempleForm = new FormGroup({});

    onSubmit() { 
        this.router.navigate(['/confirmation'], {skipLocationChange: true});
    }
}
```

> Une application coquille à la base d'une application distribuée devraient donc régir les règles d'utilisation des API du navigateur. Ce contrôle peut s'effectuer soit en encapsulant les API permis aux composants contribué ou encore en effectuant des tests automatisés qui détecteraient des interactions non permises lors de la contribution d'une composant via l'API développeur.

Reviewer: ```suggestion
Une application coquille à la base d'une application distribuée devrait donc régir les règles d'utilisation des API du navigateur. Ce contrôle peut s'effectuer soit en encapsulant les API permis aux composants contribué ou encore en effectuant des tests automatisés qui détecteraient des interactions non permises lors de la contribution d'une composant via l'API développeur.
```

#### 4.2.4 Implémentation du principe micro-frontend au modèle distribué 

La distribution de composantes *microfrontend* distribué doit être orchestrée au sein d'un point d'accès unique. Dans l'expérimentation, ce point d'accès prend la forme d'une application hybride web/mobile dite "coquille". L'application coquille offre des fonctionnalités aux services qu'elle distribue, telle que l'authentification, l'accès au navigateur et l'application des règles logicielles encadrant ceux-ci. Elle orchestre l'accès aux services en exploitant le registre qui documente leur emplacement, leur version et leur condition d'accès.

Le registre des services est un API s'adressant aux contributeurs de services et à l'application coquille. Comme son nom l'indique, il tient un registre des services rendus disponibles sur la plateforme. D'un côté, les contributeurs utilisent l'API pour inscrire et mettre à jour leurs services. De l'autre l'application coquille obtient les informations sur les services et les distribuent en conséquence.

> Le registre n'a pas été implémenté dans le cadre de l'expérimentation. L'interaction de la coquille avec celui-ci a cependant été simulée grâce à [deux fichiers de données](../app/public/assets/registre). Ceux-ci explorent la structure de donnée d'un potentiel registre et sont obtenus par la coquille lors de son démarrage. La coquille d'adapte dynamiquement aux données du registre et permet l'ajout de services sans intervention dans son code.

Reviewer: ```suggestion
Le registre n'a pas été implémenté dans le cadre de l'expérimentation. L'interaction de la coquille avec celui-ci a cependant été simulée grâce à [deux fichiers de données](../app/public/assets/registre). Ceux-ci explorent la structure de donnée d'un potentiel registre et sont obtenus par la coquille lors de son démarrage. La coquille s'adapte dynamiquement aux données du registre et permet l'ajout de services sans intervention dans son code.
```

    * Effectuer les essais fonctionnels de lecture d’écran et saisie clavier. `A11Y` 
    * Effectuer les essais exploratoires pour vérifier les goulots d'étranglement. `Perf` 
* **Une fois l’incrément de production disponible (release)** 
    * Effectuer les essais sur différents dispositifs d'assistance. `A11Y` 
    * S’arrimer avec la norme WCAG en vigueur au gouvernement. `A11Y` 
    * Valider les tests de performance de l’IU `Perf` 

## 5. Analyse des résultats 

L'hypothèse d'expérimentation avait comme objectif de vérifier si un modèle de développement distribué basé sur une approche *microfrontend* pouvait produire les conditions favorables à un service unifié d’accès à des services gouvernementaux distribué dans divers organismes. Trois marqueurs devaient démontrer l’hypothèse: 

* La démonstration du modèle avec un prototype web et un prototype mobile ; 
* Un jugement favorable des conditions d’intégration au modèle par des contributeurs potentiels; 
* Une démonstration que le modèle de développement peut se mettre à l’échelle d’une équipe produit fixe et soutenir un éventuel écosystème de contributeurs.  

> Il n’a pas été possible de présenter le modèle à d’éventuels contributeurs au moment d’écrire ce rapport. Ce marqueur n’a donc pas été vérifié. 

Author: ```suggestion
Il n’a pas été possible de présenter le modèle à d’éventuels contributeurs au moment d’écrire ce rapport. Le deuxième marqueur n’a donc pas été vérifié. 
```
